x11/misc/mathrec-1.1c/lds/lds.man
.TH LDS 6X
.SH NAME
lds \- generate and display a lattice dynamical system
.SH SYNOPSIS
.in +4n
.ti -4n
\fIlds\fR
[-2APcdmpq] [-W 
\fIwidth\fR] [-H 
\fIheight\fR] [-w
\fIwide\fR] [-h
\fIhigh\fR] [-i 
\fIinit\fR] [-n 
\fIlambda\fR ] [-b 
\fIn\fR ] [-F 
\fIfreq\fR] [-B 
\fIf | r | n\fR ]
[-C 
\fIepsilon1\fR ] [-L 
\fIepsilon2\fR ] [-R 
\fIepsilon3\fR ]
[-r 
\fIdelta\fR ] [-E 
\fIrho\fR ] [-M 
\fIomega\fR ] [-T 
\fIheight\fR] [-o 
\fIfname\fR]
.sp 2
.ft B
.SH DESCRIPTION
\fIlds\fR
generates and displays a sequence of curves or 
cellular automata like figures which graphically represent the 
evolution of a Lattice Dynamical System (LDS). An LDS is an array
of cells, each of which represents a dynamical system. Each cell
is coupled to its nearest neighbors. The dynamical systems available
in this software package are currently the logistic map, circle map
and tent map. The state of each cell in the LDS is a real number in
the unit interval. The evolution of the lattice is determined by an
iteration thru the specified dynamic (e.g. the logistic equation)
followed by a weighted averaging with its nearest neighbors. 
.sp
By default,
\fIlds\fR 
calculates and displays a 1-dimensional LDS. A 2-dimensional
LDS can be specified on the command line (see the descriptions of the
\fI-h, -2\fR and \fI-H\fR options below). In the 2-D case, nearest neighbors
are defined to be left, right, upper and lower (no diagonal neighbors).
In the 1-D case, nearest neighbors are simply left and right.
.sp
Command line arguments and run-time keyboard input allow lds to simulate
a wide variety of lattice dynamical systems. The user can specify the
dynamic to be used, the non-linearity parameter, the strength of coupling,
the initial conditions, the size of the array, the lenght of the run,
whether and how to evolve connection strengths, and more. During the run,
the display of an evolving 1-D LDS can either be CA-like with each generation
being represented as a horizontal line evolving upward in the window or as
points or a curve with X axis the lattice and Y-axis the cell states [0,1].
The display of an evolving 2-D LDS can either be CA-like with each generation
being represented as a rectangle evolving upward in the window or as points
or a curve with X axis the projected lattice and Y-axis the cell states [0,1].
In addition, graphical display of cell states can be toggled between actual
state and phase difference with left neighbor. 
.sp
Site and spatial histogram windows can also be displayed. The X-axis for
the site histogram curve is the unit interval [0,1] while the Y-axis is
the number of lattice sites which have taken on that value. The X-axis for
the spatial histogram window is the projection of the lattice down onto its
horizontal width while the Y-axis is the unit interval [0,1].
.sp
Lattice dynamical systems are also referred to as coupled map lattices.
.sp
.SH OPTIONS
.sp
.ft B
.ti 5
-2 
.ft R
.ti 8
Default to a two dimensional lattice. Defaults for width and height are now
256x256.
.ft B
.ti 5
-A 
.ft R
.ti 8
Creates site histogram curve
.ft B
.ti 5
-p 
.ft R
.ti 8
Indicates draw phase 1st differences
.ft B
.ti 5
-q 
.ft R
.ti 8
Indicates not to draw "quilt" style. In quilt mode, the lattice is moved
upward in the window each generation (if the window is larger than the
lattice). If the "-q" option is specified, subseqeunt generations overwrite 
the previous one.
.ft B
.ti 5
-c
.ft R
.ti 8
Indicates draw curves
.ft B
.ti 5
-P
.ft R
.ti 8
Indicates draw points (when in curve drawing mode)
.ft B
.ti 5
-m
.ft R
.ti 8
Indicates monochrome plot
.ft B
.ti 5
-B [ f | r | 
.IR b ]
.ft R
.in 8
Selects fixed zero, random, or specified fixed value of boundaries.
Option -Bf sets boundaries to zero; -Br selects randomly fluctuating
boundaries; and -Bb assigns value 0 < b < 1 to boundaries.
The default boundary specification is periodic.
.in -8
.ft B
.ti 5
-w
.I n
.ft R
.ti 8
Indicates the lattice is 
.I n
cells wide (default is full screen)
.ft B
.ti 5
-h
.I n
.ft R
.ti 8
Indicates the lattice is 
.I n
cells high (default is 1 in the absense of the 
.I -2
argument and 256 when 
.I -2
is specified)
.ft B
.ti 5
-W
.I n
.ft R
.ti 8
Indicates the windows are 
.I n
pixels wide (default is full screen)
.ft B
.ti 5
-H
.I n
.ft R
.ti 8
Indicates the windows are 
.I n
pixels high (default is full screen in the absense of the
.I -2
argument and 256 when
.I -2
is specified)
.ft B
.ti 5
-i [ p
.I n
| 
.I nnnn
| r ]
.ft R
.in 8
Selects initial condition. Option -ip
.I n
indicates periodic initial conditions with frequency 
.IR n .
Option -i
.I nnnnn
indicates center cells are given values 
.IR 1/n .
Option -ir selects random initial conditions (which is the default)
.in -8
.ft B
.ti 5
-n [ r | l | p
.I m
| lambda ]
.ft R
.in 8
Selects non-linearity parameter values (default is 3.7). Option -nr 
selects randomly assigned non-linearity parameters. Option -nl
selects linearly assigned non-linearity parameters. Option -np
.I m
selects periodically assigned non-linearity parameters with frequency
.IR m .
Option  -n lambda
selects non-linearity parameter lambda for all sites.
.in -8
.ft B
.ti 5
-b
.I n
.ft R
.ti 8
Begin graphing at generation 
.I n 
(default is 1)
.ft B
.ti 5
-F 
.I n
.ft R
.ti 8
Indicates display every 
.I n'th 
generation (1 is default)
.ft B
.ti 5
-C
.I epsilon1
.ft R
.ti 8
Indicates weight of center cell (default 0.9) where 0 < epsilon1 < 1.
.ft B
.ti 5
-L
.I epsilon2
.ft R
.ti 8
Indicates weight of left neighbor (default 0.05) where 0 < epsilon2 < 1.
.ft B
.ti 5
-R
.I epsilon3
.ft R
.ti 8
Indicates weight of right neighbor (default 0.05) where 0 < epsilon3 < 1.
.ft B
.ti 5
-r
.I delta
.ft R
.in 8
Indicates range outside of which differences are graphed (0 < delta < 1)
This value also serves as the determining distance over which connection
strengths weaken rather than strengthen (when the -E flag is specified).
.in -8
.ft B
.ti 5
-E
.I rho
.ft R
.in 8
Indicates the rate at which connection strengths evolve (0 < rho < 1).
Connection strengths do not change if no -E flag is present.
.in -8
.ft B
.ti 5
-M
.I omega
.ft R
.in 8
Selects circle map
.in -8
.ft B
.ti 5
-T
.I height
.ft R
.in 8
Selects tent map
.in -8
.ft B
.ti 5
-o 
.ft R
.I fname
.ti 8
Outputs graphed generations to file 
.IR fname .
.ft B
.sp
.SH NOTES
.sp
.ft R
.in 5
In the absence of either the -M or -T arguments, the logistic map is used.
.in -5
.sp
.ti 5
During display, use of the keys 
.I 123456789+-<>BDEFHILPQRSWXcdfhimpswxq? 
indicates:
.ft I
.sp
.ti 10
(1-9) Set the frequency to 1-9
.ti 10
(+) Increment the frequency by 1
.ti 10
(-) Decrement the frequency by 1
.ti 10
(>) Double the frequency
.ti 10
(<) Halve the frequency
.ti 10
(B) Begin again
.ti 10
(c) toggle curve/cell display
.ti 10
(d) draw
.ti 10
(D) Flush the drawing buffer
.ti 10
(E) Erase each generation
.ti 10
(f or F) Save the drawing window to a file
.ti 10
(h) Display histogram curve
.ti 10
(H) Histogram tracking toggled
.ti 10
(I) Increment the stripe interval
.ti 10
(i) Decrement the stripe interval
.ti 10
(L) Lines drawn
.ti 10
(m) multi-step
.ti 10
(p) Toggle display of phase 1st differences
.ti 10
(P) Points drawn
.ti 10
(R) Spin the color wheel
.ti 10
(s) single step 
.ti 10
(S) Spin the color wheel and increment the spin length
.ti 10
(w) Decrement the color wheel index
.ti 10
(W) Increment the color wheel index
.ti 10
(x) Clear the window
.ti 10
(X) Toggle complex dynamical systems mode
.ti 10
(Q or q) quit
.sp 2
.ft B
.SH AUTHOR
.sp
.ft R
.in 5
.I lds
was written by Ronald Record. Questions, suggestions, and comments may be
directed via e-mail to rr@sco.com or ...uunet!sco!rr.
.sp 2
.ft B
.SH REFERENCES
.sp
.ft R
The literature is sparse but includes the following excellent papers :
.sp
"Lyapunov Analysis and Information Flow in Coupled Map Lattices" by
Kunihiko Kaneko, Physica 23D (1986) 436-447
.sp
"Spatiotemporal Chaos in One- and Two- Dimensional Coupled Map Lattices"
by Kaneko
.sp
"Spatiotemporal Chaos and Noise" by Gottfried Mayer-Kress and Kaneko,
J. Stat. Phys.(1988)
.sp
"Pattern Competition Intermittency and Selective Flicker Noise in
Spatiotemporal Chaos" by Kaneko, Physics Letters A, V125, 1 (1987)
.sp
"Pattern Dynamics in Spatiotemporal Chaos" by Kaneko (1987)
.sp
"Phenomenology of Spatial-Temporal Chaos" by Jim Crutchfield and Kaneko,
appearing as a chapter in "Directions in Chaos" edited by Hao Bai-lin,
World Scientific Publishing (1987)
.sp
"Robust Space-Time Intermittency and 1/f Noise" by James Keeler and Doyne
Farmer, Physica 23D (1986) 413-435
.sp


x11/misc/mathrec-1.1/lds/lds.c
/*************************************************************************
 *                                                                       *
 *  Copyright (c) 1988-1993 Ronald Joe Record                           *
 *                                                                       *
 *  All rights reserved. No part of this program or publication may be   *
 *  reproduced, transmitted, transcribed, stored in a retrieval system,  *
 *  or translated into any language or computer language, in any form or *
 *  by any means, electronic, mechanical, magnetic, optical, chemical,   *
 *  biological, or otherwise, without the prior written permission of:   *
 *                                                                       *
 *      Ronald Joe Record (408) 458-3718                                 *
 *      212 Owen St., Santa Cruz, California 95062 USA                   *
 *                                                                       *
 *************************************************************************/

/*
 *	Written 2/22/88 by Ronald Record (sco!rr) in partial fulfillment of
 *	the Mathematics Ph.D. requirements at the University of California at 
 *	Santa Cruz. This work is in part based upon articles by Doyne Farmer, 
 *	Jim Crutchfield and Kunihiko Kaneko.
 *
 *  X11 port begun 20 Jan 93 by Ronald Record (rr@sco.com) for fun
 */

#include <math.h>
#include <memory.h>
#include <stdio.h>
#include <stdlib.h>
#include <values.h>
#include "x.h"
#include "defines.h"
#include "lds.h"

/* routines in this file */
void usage(), Clear();
void redisplay(), BufferSpace(), FlushHisbuf();
void resize(), event_loop(), freemem(), zero_mem();
void spahist(), hist(), adjconn(), zerohist(), zerospa();
int complds(), upd_lambda();
double map();

/* external routines called in this file */
extern void Getkey(), draw_hist(), init_mem(), init_gen(), next_gen();
extern void InitBuffer();

void
usage()
{
  printf("\nUsage: lds [-PSpAcdx] [-w wide] [-h high] [-n lambda] [-b#]\n");
  printf("\t\t[-f#] [-F freq] [-C EC] [-L EL] [-R ER] [-U EU] [-D ED]\n");
  printf("\t\t[-W width] [-H height] [-r range] [-M omega] [-T tall] [-E rate] [-i init]\n");
	printf("Where :\n");
	printf("\t'-A' creates site histogram curve\n");
	printf("\t'-p' indicates draw phase 1st differences\n");
	printf("\t'-c' indicates draw curves\n");
	printf("\t'-d' indicates demo mode (use with -f# option)\n");
	printf("\t'-x' indicates complex dynamical system mode\n");
	printf("\t'-P' indicates draw points (default is curves)\n");
	printf("\t'-[M|T] #' indicates use circle|tent map with omega|height=#\n");
	printf("\t'-E #' indicates evolve connection strengths at rate #\n");
	printf("\tEC|EL|ER|EU|ED indicates weight of ctr|lft|rgt|upr|lwr cell\n");
	printf("\twidth\t= width of window (default is full screen)\n");
	printf("\theight\t= height of window (default is full screen)\n");
	printf("\twide\t= width of lattice (default is full screen)\n");
	printf("\thigh\t= height of lattice (default is full screen)\n");
	printf("\tinit\t= initial condition (default is random)\n");
	printf("\tlambda\t= logistic parameter (default is 3.7)\n");
	printf("\t-b#\t= first generation graphed (default is 1)\n");
	printf("\t-f#\t= Final generation graphed (default is MAXSHORT)\n");
	printf("\tfreq\t= display every freq'th generation (1 is default)\n");
	printf("\trange\t= range outside of which differences are graphed\n");
	printf("During display, use of the keys 'cseEdBmHPLSXq' indicates:\n");
	printf("'continue' 'single step' 'erase' 'Erase all' 'draw' 'Begin again' 'multi-step'\n");
	printf("'Histogram' 'Points' 'Lines' 'Spin one color' 'Xtended spin' 'quit'\n");
		exit(1);
}

void
Clear(w) 
Window w;
{
	if (w == canvas) {
		XFillRectangle(dpy, pixmap, Data_GC[0], 0, 0, width, height);
		XCopyArea(dpy, pixmap, w, Data_GC[0], 0, 0, width, height, 0, 0);
		InitBuffer(&Points, numcolors);
	}
	else if (w == spahis) {
		XFillRectangle(dpy, spamap, Data_GC[0], 0, 0, width, height);
		XCopyArea(dpy, spamap, w, Data_GC[0], 0, 0, width, height, 0, 0);
		InitBuffer(&Points, numcolors);
	}
	else
		XClearWindow(dpy, w);
}

void
BufferSpace(color, x, y)
int color;
int x, y;
{
	if (Hpoints.npoints[color] == MAXPOINTS)
	{
		XDrawPoints(dpy, spahis, Data_GC[color],
		    Hpoints.data[color], Hpoints.npoints[color], CoordModeOrigin);
		XDrawPoints(dpy, spamap, Data_GC[color],
		    Hpoints.data[color], Hpoints.npoints[color], CoordModeOrigin);
		Hpoints.npoints[color] = 0;
	}
	Hpoints.data[color][Hpoints.npoints[color]].x = x;
	Hpoints.data[color][Hpoints.npoints[color]].y = height - y;
	++Hpoints.npoints[color];
}

void
FlushHisbuf()
{
	int color;
	for (color = mincolor; color < numcolors; ++color)
		if (Hpoints.npoints[color])
		{
		    XDrawPoints(dpy, spahis, Data_GC[color],
			    Hpoints.data[color], Hpoints.npoints[color], CoordModeOrigin);
		    XDrawPoints(dpy, spamap, Data_GC[color],
			    Hpoints.data[color], Hpoints.npoints[color], CoordModeOrigin);
		    Hpoints.npoints[color] = 0;
		}
}

void
redisplay(ev, w)
XExposeEvent    *ev;
Window w;
{
	/*
 	* Extract the exposed area from the event and copy
 	* from the saved pixmap to the window.
 	*/
	if (w == canvas)
		XCopyArea(dpy, pixmap, canvas, Data_GC[0], 
           ev->x, ev->y, ev->width, ev->height, ev->x, ev->y);
	else if (w == spahis)
		XCopyArea(dpy, spamap, spahis, Data_GC[0], 
           ev->x, ev->y, ev->width, ev->height, ev->x, ev->y);
}

void
resize(w)
Window w;
{
	Window r;
	int x, y;
	unsigned int bw, d, new_w, new_h;
	static XWindowChanges values;

	XGetGeometry(dpy,w,&r,&x,&y,&new_w,&new_h,&bw,&d);
	if ((new_w == width) && (new_h == height))
		return;
	freemem();
	width = new_w; height = new_h;
	XClearWindow(dpy, canvas);
	if (sflag)
		XClearWindow(dpy, spahis);
	if (hflag)
		XClearWindow(dpy, hiswin);
	if (pixmap)
		XFreePixmap(dpy, pixmap);
	pixmap = XCreatePixmap(dpy, DefaultRootWindow(dpy), 
			width, height, DefaultDepth(dpy, screen));
	values.width = width;
	values.height = height;
	if (w == canvas) {
		XConfigureWindow(dpy,hiswin,CWWidth|CWHeight,&values);
		XConfigureWindow(dpy,spahis,CWWidth|CWHeight,&values);
	}
	else if (w == hiswin) {
		XConfigureWindow(dpy,canvas,CWWidth|CWHeight,&values);
		XConfigureWindow(dpy,spahis,CWWidth|CWHeight,&values);
	}
	else {
		XConfigureWindow(dpy,canvas,CWWidth|CWHeight,&values);
		XConfigureWindow(dpy,hiswin,CWWidth|CWHeight,&values);
	}
	xpoint = ypoint = 0;
	run = 1;
	init_mem();
	InitBuffer(&Points, numcolors);
	Clear(canvas);
	init_gen();
}

int
complds()
{
	if (!run)
		return(TRUE);
	next_gen();
	if (hflag)
		draw_hist();
	return(FALSE);
}

void
event_loop()
{
	int n;
	XEvent event;

	if (complds() == TRUE)
	    run=0;
	n = XEventsQueued(dpy, QueuedAfterFlush);
	while (n--) {
        XNextEvent(dpy, &event);
        switch(event.type) 
        {
        case KeyPress:
			Getkey(&event);
			break;
        case Expose:
			redisplay(&event, event.xexpose.window);
	       	break;
        case ConfigureNotify:
			resize(event.xconfigure.window);
	      	break;
        }
    }
}

void
freemem()
{
	static int i;
	extern XPoint *hpoints;

	for (i=0; i<width; i++)
		free(histarray[i]);
	for (i=0; i<high; i++) {
		free(lambda[i]);
		free(lftconn[i]);
		free(rgtconn[i]);
		free(ctrconn[i]);
		free(uprconn[i]);
		free(lwrconn[i]);
		free(currentgen[i]);
		free(nextgen[i]);
		free(avg[i]);
	}
	free(lambda);
	free(lftconn);
	free(rgtconn);
	free(ctrconn);
	free(uprconn);
	free(lwrconn);
	free(currentgen);
	free(nextgen);
	free(histogram);
	free(histarray);
	free(diff); 
	free(avg);
	free(hpoints);
}

void
zerospa()
{
	static int i, j;

	for (i=0; i<width; i++)
		for (j=0; j<height; j++)
			histarray[i][j] = 0;
}

void
zerohist()
{
	static int i;

	for (i=0; i<width; i++)
		histogram[i] = 0;
}

void
zero_mem()
{
	static int i, j;

	for (i=0; i<wide; i++) {
		diff[i] = 0.0;
		for (j=0; j<high; j++) {
			lambda[j][i] = 0.0;
			lftconn[j][i] = 0.0;
			rgtconn[j][i] = 0.0;
			ctrconn[j][i] = 0.0;
			currentgen[j][i] = 0.0;
			nextgen[j][i] = 0.0;
			avg[j][i] = 0.0;
		}
		zerohist();
	}
}

int
upd_lambda() 
{
	static int i, j;

	if (xavg)
	  if (numgen % xavg) {
		for (i=0;i<wide;i++)
		  for (j=0;j<high;j++)
			avg[j][i] += currentgen[j][i];
		return(1);
	  }
	for (i=0;i<wide;i++)
	  for (j=0;j<high;j++)
		avg[j][i] = (avg[j][i] + (double)currentgen[j][i]) / (double)xavg;
	for (i=0;i<high;i++) {
	  if (circle)
		lambda[i][0] = avg[i][wide-1]/M_2PI;
	  else if (tent)
		lambda[i][0] = avg[i][wide-1];
	  else 
		lambda[i][0] = 3.0 + avg[i][wide-1];
	}
	for (i=1;i<wide;i++)
	  for (j=0;j<high;j++) {
		if (circle)
			lambda[j][i] = avg[j][i-1]/M_2PI;
		else if (tent)
			lambda[j][i] = avg[j][i-1];
		else 
			lambda[j][i] = 3.0 + avg[j][i-1];
	  }
	return(0);
}

void
spahist(sites)
double *sites;
{
	static int i, j, k, l, n;

	for (i=0;i<wide;i++) {
		j = (int)(sites[i]*height);
		k = (i*width)/wide;
		n = ++histarray[k][j];
		if (n > maxhist)
			maxhist = n;
		if (n < minhist)
			minhist = n;
		l = ((histarray[k][j]-minhist)*(numcolors-mincolor-1)/maxhist)+mincolor;
		BufferSpace(l, k, j);
	}
}

void
hist(sites)
double *sites;
{
	int i;

	for (i=0;i<wide;i++)
		histogram[(int)floor(sites[i]*width)]++;
}

void
adjconn()
{
	static int i,j,k,l,m,n;

	for (i=0;i<wide;i++) {
		if (i == wide -1)
			k = 0;
		else
			k = i + 1;
		if (i == 0)
			l = wide - 1;
		else
			l = i - 1;
		for (j=0;j<high;j++) {
			if (j == high -1)
				m = 0;
			else
				m = j + 1;
			if (j == 0)
				n = high - 1;
			else
				n = j - 1;
			rgtconn[j][i]=erate*(currentgen[j][i]*currentgen[j][k]);
			lftconn[j][i]=erate*(currentgen[j][i]*currentgen[j][l]);
			uprconn[j][i]=erate*(currentgen[j][i]*currentgen[m][i]);
			lwrconn[j][i]=erate*(currentgen[j][i]*currentgen[n][i]);
			ctrconn[j][i] = 1.0 - 
				(rgtconn[j][i] + lftconn[j][i] + uprconn[j][i] + lwrconn[j][i]);
		}
	}
}

double
map(npar,x)
double npar;
double x;
{
	double y;

	x = fmod(x,1.0);	/*	Make sure value is between 0 and 1 */
	if (logistic)
		y = npar * x * (1.0 - x);
	else if (circle)
		y = omega + x - npar * sin(M_2PI*x);
	else if (tent)
		if (x < npar)
			y = incline*x;
		else
			y = decline*(x-1.0);
	y = fmod(y,1.0);	/*	Make sure return value is between 0 and 1 */
	return(y);
}
